fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -0,0 +1,1143 @@
+		std::shared_ptr<KmsUrlCtx> curUrl = ctx->kmsUrlHeap.top();
+
+		TraceEvent("DropCachedKmsUrls", ctx->uid)
+		    .detail("Url", Traceable<std::string>::toString(curUrl->url))
+		    .detail("NumRequests", curUrl->nRequests)
+		    .detail("NumFailedResponses", curUrl->nFailedResponses)
+		    .detail("NumRespParseFailures", curUrl->nResponseParseFailures);
+
+		ctx->kmsUrlHeap.pop();
+	}
+}
+
+bool shouldRefreshKmsUrls(Reference<RESTKmsConnectorCtx> ctx) {
+	if (!FLOW_KNOBS->REST_KMS_CONNECTOR_REFRESH_KMS_URLS) {
+		return false;
+	}
+
+	return (now() - ctx->lastKmsUrlsRefreshTs) > FLOW_KNOBS->REST_KMS_CONNECTOR_REFRESH_KMS_URLS_INTERVAL_SEC;
+}
+
+void extractKmsUrls(Reference<RESTKmsConnectorCtx> ctx, rapidjson::Document& doc, Reference<HTTP::Response> httpResp) {
+	// Refresh KmsUrls cache
+	dropCachedKmsUrls(ctx);
+	ASSERT(ctx->kmsUrlHeap.empty());
+
+	for (const auto& url : doc[KMS_URLS_TAG].GetArray()) {
+		if (!url.IsString()) {
+			TraceEvent("DiscoverKmsUrls_MalformedResp", ctx->uid)
+			    .detail("ResponseContent", Traceable<std::string>::toString(httpResp->content));
+			throw operation_failed();
+		}
+
+		std::string urlStr;
+		urlStr.resize(url.GetStringLength());
+		memcpy(urlStr.data(), url.GetString(), url.GetStringLength());
+
+		TraceEvent("DiscoverKmsUrls_AddUrl", ctx->uid).detail("Url", Traceable<std::string>::toString(urlStr));
+
+		ctx->kmsUrlHeap.emplace(std::make_shared<KmsUrlCtx>(urlStr));
+	}
+
+	// Update Kms URLs refresh timestamp
+	ctx->lastKmsUrlsRefreshTs = now();
+}
+
+void parseDiscoverKmsUrlsResp(Reference<RESTKmsConnectorCtx> ctx, Reference<HTTP::Response> resp) {
+	ASSERT_EQ(resp->code, HTTP::HTTP_STATUS_CODE_OK);
+
+	// Acceptable REST JSON response schema:
+	//
+	// response_json_payload {
+	//   "kmsUrls" : [ url1, url2, ...]
+	// }
+
+	TraceEvent("ParseDiscoverKmsUrls_Response", ctx->uid)
+	    .detail("RespCode", resp->code)
+	    .detail("RespContent", Traceable<std::string>::toString(resp->content));
+
+	rapidjson::Document doc;
+	doc.Parse(resp->content.c_str());
+	if (!doc.HasMember(KMS_URLS_TAG) || !doc[KMS_URLS_TAG].IsArray()) {
+		TraceEvent("DiscoverKmsUrls_MalformedResp", ctx->uid)
+		    .detail("ResponseContent", Traceable<std::string>::toString(resp->content));
+		throw operation_failed();
+	}
+
+	extractKmsUrls(ctx, doc, resp);
+}
+
+ACTOR Future<Void> discoverKmsUrls(Reference<RESTKmsConnectorCtx> ctx) {
+	StringRef kmsDiscoveryUrls(FLOW_KNOBS->REST_KMS_CONNECTOR_KMS_DISCOVERY_URLS);
+	state std::vector<std::string> urls;
+
+	while (kmsDiscoveryUrls.empty()) {
+		StringRef u = kmsDiscoveryUrls.eat(",");
+		urls.push_back(u.toString());
+	}
+
+	if (urls.empty()) {
+		TraceEvent("DiscoverKmsUrls_Empty", ctx->uid).log();
+		throw operation_failed();
+	}
+
+	state Reference<HTTP::Response> resp;
+	state bool done = false;
+	state int i = 0;
+	for (; i < urls.size(); i++) {
+		try {
+			TraceEvent("DiscoverKmsUrls", ctx->uid).detail("Url", Traceable<std::string>::toString(urls[i]));
```

Comment:
This is not how `Traceable` should be used. If `urls[[i]` is of type `std::string` you can just pass it to `TraceEvent::detail`. 

Maybe let me give you an example of how this works. Let's say you have a type `MyType` which implements `std::string MyType::toString()`. Now you could do something like `TraceEvent(SevDebug, "Foo:).detail("Bar", myTypeInstance.toString());` But this has two drawbacks:

1. It's a bit ugly
2. When we run in prod, `MyType::toString()` will still be called but the result is not used. So this is wasted work.

To solve both of these problems, you can specialize `Traceable`. This would look like this:

```
template <>
struct Traceable<MyType> : std::true_type {
	static std::string toString(const MyType& value) { return value.toString(); }
};
```

Now you can change above code to do this:

```
TraceEvent(SevDebug, "Foo:).detail("Bar", myTypeInstance);
```

So this will be nicer looking (no need to call `MyType::toString()` anymore) and therefore addresses the first problem. It also addresses the second problem, because `TraceEvent` now will not call `Traceable<MyType>::toString` unless the TraceEvent is actually enabled.

Hope this makes sense.

Reply:
Thanks for detailed explanation, Markus.